Simple plugin system for Gaia

mbertrand opened the pull request:
Uses pkg_resources.iter_entry_points to load any installed Gaia plugin modules and their configuration settings.

A sample plugin demo project with instructions is at https://github.com/mbertrand/gaia_plugin_demo

Addresses issue https://github.com/OpenDataAnalytics/gaia/issues/69

aashish24 commented:
This is a nice / must have feature @mbertrand thanks for pushing it.. I will have a look at it today / tomorrow. It would be nice if others can provide their feedback as well @jbeezley @kotfic @cjh1 @dorukozturk

aashish24 commented on gaia/parser.py:
```diff
@@ -35,12 +40,29 @@
     gaia.geo.geo_inputs, inspect.isclass) if x[0].endswith('IO')])
 
 
+def load_plugins():
```
@mbertrand I think there might be a better place for this function. Not entirely sure if parser is the right place for it. Thoughts?

aashish24 replied:
Also, should we make geo as plugin as part of this PR?

mbertrand replied:
@aashish24 I can move the load_plugins() definition to gaia/__init__.py or gaia/core.py, do you have a preference?

The raster and vector parts of geo seem more like core functionality, but I think some of it could be split up into multiple plugins, for instance:
  - twitter plugin
  - stats plugin (anything that depends on the pysal module) 
  - maybe a raster plugin too, but keep the vector io's/processes in geo as core functionality?

aashish24 replied:
> @aashish24 I can move the load_plugins() definition to gaia/init.py or gaia/core.py, do you have a preference?

core.py sounds better to me. 



> twitter plugin
> stats plugin (anything that depends on the pysal module)
> maybe a raster plugin too, but keep the vector io's/processes in geo as core functionality?

+1, that's what I was thinking. May be we can keep the base classes for raster and vector in core and move the actual implementations as plugins?

mbertrand replied:
@aashish24 I moved the function to core.py

My preference would be to keep at minimum the basic vector inputs and processes included in gaia, providing some built-in functionality without requiring the installation of an additional plugin. 

Maybe after this is merged into master we can create a repo for each plugin and move the relevant code over?

aashish24 replied:
> Maybe after this is merged into master we can create a repo for each plugin and move the relevant code over?

+1 for this

aashish24 replied:
> My preference would be to keep at minimum the basic vector inputs and processes included in gaia, providing some built-in functionality without requiring the installation of an additional plugin.

Sounds good to me for now.

aashish24 commented:
@mbertrand should I look at this again?

mbertrand commented:
@aashish24 sure, I think it's ready to go

aashish24 commented:
@mbertrand this looks good but we are missing documentation and how we a user/developer can add a new plugin. May be add that in the next PR?

LGTM :+1:

mbertrand commented:
@aashish24 that sounds good to me.  Once some of the existing code is moved over to new plugin repos they can be used as examples in documentation.

kotfic commented on gaia/core.py:
```diff
@@ -66,4 +69,18 @@ def get_config(config_file=None):
+        importlib.import_module(module.__name__)
+        installed_plugins.append(module)
+        plugin_config = get_config(
+            config_file='{}/gaia.cfg'.format(os.path.dirname(module.__file__)))
```
@mbertrand Can we do this through a known function defined inside the module? e.g.
```python
if hasattr(module, 'get_config'): 
    config.update(module.get_config())
```
That way the core gaia code doesn't enforce a particular config mechanism on the plugins  (e.g. gaia.cfg in root directory).

kotfic commented on gaia/parser.py:
```diff
@@ -33,6 +36,9 @@
 valid_classes.extend([x[0] for x in inspect.getmembers(
     gaia.geo.geo_inputs, inspect.isclass) if x[0].endswith('IO')])
+for plugin in get_plugins():
+    valid_classes.extend([x[0] for x in inspect.getmembers(
```
@mbertrand Rather than add all classes from a plugin,  can we add something to the plugin API like ```PLUGIN_CLASS_EXPORTS``` which must be present on the plugin and explicitly defines what classes should be included?

mbertrand replied:
Thanks @kotfic for the great suggestions, I'll get started on them.

aashish24 replied:
nice idea @kotfic :+1: :+1:

mbertrand commented:
@kotfic I incorporated your suggestions if you want to look them over.  @aashish24 I also added preliminary documentation for plugins at https://github.com/OpenDataAnalytics/gaia/blob/90208d855599fc2c8f5fd7bf02865e50802f708b/docs/plugins.rst

kotfic commented:
I would wait on @aashish24's review,  but this looks good to me.

aashish24 commented:
> https://github.com/OpenDataAnalytics/gaia/blob/90208d855599fc2c8f5fd7bf02865e50802f708b/docs/plugins.rst

Thanks @mbertrand looks good

+1 LGTM
